### DISCUSSION TITLE
Removing async_trait dependency as for Rust 1.75

### DISCUSSION THREAD

**Discussion by JasterV** [bikeshedder/deadpool#294] [Ideas]
Hello there!

Since[ Rust 1.75](https://releases.rs/docs/1.75.0/) was released and `async` traits are now supported natively, there is no need to keep relying on the [`async_trait`](https://crates.io/crates/async-trait) crate.

How hard would it be to get rid of this dependency? Are there any plans for it yet?

**Comment by bikeshedder** [comment]
I have asked that same question myself. At the time of this writing the Rust compiler warns against using `async fn` for public traits. It could be changed into `-> impl Future<...>` though I have not tried that, yet:

https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html#dynamic-dispatch

> Traits that use `-> impl Trait` and `async fn` are not object-safe, which means they lack support for dynamic dispatch. We plan to provide utilities that enable dynamic dispatch in an upcoming version of the `trait-variant` crate.

In theory this should be no limiting factor as the `Manager` is a generic and there should be no dynamic dispatch happening.

I need to look into that. I guess there are dragons lurking. :dragon:

**Reply by Johnabell** [reply]
> I guess there are dragons lurking. 🐉

There are definitely considerations regarding the use of this feature in public traits in particular this write up is a good summary: https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html#where-the-gaps-lie. Moving away from `async_trait` might change the implicit bounds that are applied to the return types and consequently the returned `Future` might no longer be `Send` which could be problematic for multi-threaded async runtimes (and represent a breaking change). Additionally, the fact that the traits are no longer considered object safe should also be considered a breaking change.

**Comment by bikeshedder** [comment]
I just gave this a try:
- #297 

I had to change some detail in the `deadpool-postgres` crate as the swappable TLS implementation caused some issues. I solved it via cloning the config structure inside the `Connect` trait I provided to make this object safe.

**Comment by bikeshedder** [comment]
I just merged that PR. All tests passed and it seams to be working just fine.

If I don't encounter any blocker I will release the next deadpool version without `async_trait`. :tada:

**Reply by JasterV** [reply]
Damn that is so cool! I'm glad that my comment started all this, thank you for your blazingly fast response!

**Reply by JasterV** [reply]
Also thank you for updating the edition of the crate to the latest 2021 and bumping dependencies :D
